32-tgyuuAn

tgyuuAn opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
[멀티탭 스케줄링](https://www.acmicpc.net/problem/1700)

## ✔️ 소요된 시간
**30분**
<br><br><br>

내가 약한 문제 부류 **그리디**를 공부하기 위해서 백준 그리디 문제를 찾아나섰다.

그 중 골드 문제 아무거나 집어서 풀었는데, [카드 정렬하기](https://www.acmicpc.net/problem/1715) 라는 문제를 원래 먼저 풀었었는데,

아니 이 문제 왜 골4인가 싶다. **진짜 거짓말 안치고 8분만에 풀었다**.

점수 올리고 싶으시면 저 문제 추천드리리라.

차마 저 문제로 PR을 올리기엔 뭐 해서 그리디 다른 문제를 또 풀어보았다.
 
<br>

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/116813010/97624606-839e-4099-95a4-b524dfd1db87)
## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->

**그리디 알고리즘**의 **핵심**은 

무턱대고 바로 코드를 짜는 것이 아니라,

**진짜 이 문제를 풀기 위해서 기가 막히고 적합한 아이디어가 떠오를 때 까지 키보드에서 잠시 손을 내려놓고 생각만 하는 것**이 핵심이다. 

_(사실 어떤 알고리즘이든 그렇긴하다)_

<br><br>

이 문제는 약간 **운영체제나 컴퓨터 구조**에서 배운 **캐싱 스케줄링 LRU나 LFU** 같아 보였다.

뭔가 이런 식으로 접근 하면 될 것 같아서 여기서 약간 생각을 확장해나갔다.

<br><br><br><br><br>

### 생각의 흐름 🤔💭

<br>

>플러그 구멍이 N개라고 했을 때,
>
>만약 N개가 가득 찼다.
>
>그러면 다음 거를 파악함.

<br><br>

>이 때, 지금 꽂혀있는 플러그의 번호들이 나올 때 까지 계속 탐색함.
>
>ex) 2,3이 꽂혀있다면 다음 번 2가 나올 때 까지, 다음 번 3이 나올때 까지 각각 탐색
>
>만약 다음 번에 해당 플러그가 없을 경우 바로 뽑아도 됌

<br><br>

>ex) 구멍 = 2개,
>
>주어진 가전기기 사용 = 2, 3, 4, 5, 2, 3 일 때,
>
>일단,
>
>뒤에 해당 플러그가 없으면 1순위로 뽑아도 되는 것
>
>그리고 각각 몇번 째 뒤에 나오는 지 파악하고 뽑자.
>
>이거네...!!

<br><br>

진짜 이 생각이 바로 떠올랐고,

이렇게 코드를 짜서 제출해보니 진짜 정답이었다.

**개꿀**

<br><br><br>

코드에 꽂혀있는 전자기기 플러그의 그 다음 순번을 찾기가 귀찮아서 `index`를 이용하고 `try` `except` 해주었다.

만약 에러가 날 경우 뒤에 재사용 하지 않는 것이므로 코스트를 최대치로 올려주었다.

<br>

```python
from collections import defaultdict
from heapq import *

socket_count, use_count = map(int,input().split())
use_list = list(map(int,input().split()))

now_socket = []
answer = 0
for idx, use in enumerate(use_list):
    # print(now_socket, use)
    if use in now_socket:
        continue
    
    if len(now_socket) < socket_count:
        now_socket.append(use)
        continue

    else:
        candidate = []
        for check_idx, using in enumerate(now_socket):
            try:
                next_idx = use_list[idx:].index(using)
            
            # 찾지 못했을 경우 코스트를 최대로 올림
            except:
                next_idx = 101

            heappush(candidate,[-next_idx,check_idx])

        next_idx, remove_idx = heappop(candidate)

        now_socket[remove_idx] = use
        answer += 1

print(answer)
```

## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

당분간 **DP**랑 **그리디, 재귀** 좀 조져야겠다.

pknujsp reviewed (approved):
처음 풀 때 우선적으로 **N**개 포트에 꽂을 수 있는거 순서대로 다 꽂고나서, 그 이후 부터는 가장 덜 쓰는 거 순으로 뽑아나가게 풀었다가 바로 틀렸네요.

다시 곰곰이 생각해보니 순서대로 멀티탭에 꽂아보다가 멀티탭에 자리가 없으면 제일 나중에 쓰는거를 뽑아가면서 진행하면 되더라구요.

문제 풀 때 생각의 방향을 어떻게 해야할지 좀 더 잡아주는 좋은 문제였습니다

```python
from sys import *
from heapq import *
from collections import *

N, K = map(int, stdin.readline().split())
ARR = list(map(int, stdin.readline().split()))

groups = {}
for i, v in enumerate(ARR):
    groups.setdefault(v, deque()).append(i)

tabs = set()
tabq = []
count = 0
skipped_devices = set()

for device in ARR:
    groups[device].popleft()

    if not groups[device]:
        groups[device].append(1000)

    if device in tabs:
        skipped_devices .add(device)
        continue

    if len(tabs) < N:
        heappush(tabq, (-groups[device][0], device))
        tabs.add(device)
        continue

    if skipped_devices :
        for i in range(len(tabq)):
            idx, q_device = tabq[i]

            if idx != -groups[q_device][0]:
                first_idx = -groups[q_device][0]
                tabq[i] = (first_idx, q_device)
                skipped_devices .remove(q_device)

                if not skipped_devices :
                    break
        heapify(tabq)

    tabs.remove(heappop(tabq)[1])
    tabs.add(device)
    heappush(tabq, (-groups[device][0], device))
    count += 1

print(count)
```

최소 힙으로 구현했고 순회하다가 멀티탭에 이미 기기가 꽂혀있으면 최소 힙에 있는 기기의 다음 사용 순서가 틀어져서
재 정렬하게 로직 달았습니다

pknujsp commented on tgyuuAn/그리디/멀티탭 스케줄링.py:
```diff
@@ -0,0 +1,35 @@
+
+    else:
+        candidate = []
+        for check_idx, using in enumerate(now_socket):
```
다음 인덱스가 가장 먼 기기를 찾을 떄 힙을 보정하는 것 대신
반복문으로 찾아도 되는 방법이 있었네요
이 부분은 생각 못했네요

pknujsp commented on tgyuuAn/그리디/멀티탭 스케줄링.py:
```diff
@@ -0,0 +1,35 @@
+
+    else:
+        candidate = []
+        for check_idx, using in enumerate(now_socket):
+            try:
+                next_idx = use_list[idx:].index(using)
+            
+            # 찾지 못했을 경우 코스트를 최대로 올림
+            except:
+                next_idx = 101
+
+            heappush(candidate,[-next_idx,check_idx])
```
생각 못했던건데 가장 먼 인덱스를 찾아낼 때 제 구현인 힙을 보정하는 것 대신에
반복문으로 간단하게 찾아내는 방법이 있었네요
K가 최대 100이니 시간복잡도도 그렇게 안 크군요

yuna83 commented on tgyuuAn/그리디/멀티탭 스케줄링.py:
```diff
@@ -0,0 +1,35 @@
+        now_socket.append(use)
+        continue
+
+    else:
+        candidate = []
+        for check_idx, using in enumerate(now_socket):
+            try:
+                next_idx = use_list[idx:].index(using)
+            
+            # 찾지 못했을 경우 코스트를 최대로 올림
+            except:
+                next_idx = 101
```
태규님 생각의 흐름을 따라가니 풀이가 딱 이해되네요
어떤 플러그를 뽑을 지 결정하는 코드를 어떻게 전개해야 하나 싶었는데
```next_idx = use_list[idx:].index(using)```
이런식으로 지금 꽂혀있는 콘센트가 다시 쓰이는 지 확인하면 되는군요!!
사용이 안된다면 뽑는 걸 **비용을 최대로 한다**는 방법도  그렇고 배울 게 넘 많은 풀이네요
많이 배워갑니다!!👍

tgyuuAn replied:
그리디 같은 문제는 유나님도 충분히 풀 수 있다고 생각합니다!

만약에 PR에 그리디 종류의 문제가 나오면 유나님도 한번 도전해봐요!!!!

tgyuuAn commented:
> 처음 풀 때 우선적으로 **N**개 포트에 꽂을 수 있는거 순서대로 다 꽂고나서, 그 이후 부터는 가장 덜 쓰는 거 순으로 뽑아나가게 풀었다가 바로 틀렸네요.
> 
> 다시 곰곰이 생각해보니 순서대로 멀티탭에 꽂아보다가 멀티탭에 자리가 없으면 제일 나중에 쓰는거를 뽑아가면서 진행하면 되더라구요.
> 
> 문제 풀 때 생각의 방향을 어떻게 해야할지 좀 더 잡아주는 좋은 문제였습니다
> 
> ```python
> from sys import *
> from heapq import *
> from collections import *
> 
> N, K = map(int, stdin.readline().split())
> ARR = list(map(int, stdin.readline().split()))
> 
> groups = {}
> for i, v in enumerate(ARR):
>     groups.setdefault(v, deque()).append(i)
> 
> tabs = set()
> tabq = []
> count = 0
> skipped_devices = set()
> 
> for device in ARR:
>     groups[device].popleft()
> 
>     if not groups[device]:
>         groups[device].append(1000)
> 
>     if device in tabs:
>         skipped_devices .add(device)
>         continue
> 
>     if len(tabs) < N:
>         heappush(tabq, (-groups[device][0], device))
>         tabs.add(device)
>         continue
> 
>     if skipped_devices :
>         for i in range(len(tabq)):
>             idx, q_device = tabq[i]
> 
>             if idx != -groups[q_device][0]:
>                 first_idx = -groups[q_device][0]
>                 tabq[i] = (first_idx, q_device)
>                 skipped_devices .remove(q_device)
> 
>                 if not skipped_devices :
>                     break
>         heapify(tabq)
> 
>     tabs.remove(heappop(tabq)[1])
>     tabs.add(device)
>     heappush(tabq, (-groups[device][0], device))
>     count += 1
> 
> print(count)
> ```
> 
> 최소 힙으로 구현했고 순회하다가 멀티탭에 이미 기기가 꽂혀있으면 최소 힙에 있는 기기의 다음 사용 순서가 틀어져서 재 정렬하게 로직 달았습니다

**그 이후 부터는 가장 덜 쓰는 거 순으로 뽑아나가게 풀었다가 바로 틀렸네요.** 

완전 컴퓨터 구존가 운영체제에서 배웠던 캐싱에서 제거할 때 쓰는 방법으로 하셨군요

저도 약간 그런 느낌으로 사고하다 보니 술술 풀리더라고요 .... 역시 CS 중요혀

Munbin-Lee commented:
<details>
<summary>Code</summary>

```cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, K;
    cin >> N >> K;

    vector<int> items(K);

    for (int &item: items) {
        cin >> item;
    }

    vector<queue<int>> usages(K + 1);

    for (int time = 0; time < K; time++) {
        int item = items[time];
        usages[item].emplace(time);
    }

    const int INF = 987654321;

    auto getNextUsageTime = [&](int item, int curTime) {
        auto &usage = usages[item];

        while (!usage.empty()) {
            int usageTime = usage.front();

            if (usageTime <= curTime) {
                usage.pop();
                continue;
            }

            return usageTime;
        }

        return INF;
    };

    unordered_set<int> connected;

    auto getItemToRemove = [&](int curTime) {
        int maxNextUsageTime = -1;
        int itemToRemove = -1;

        for (auto item: connected) {
            int nextUsageTime = getNextUsageTime(item, curTime);

            if (nextUsageTime <= maxNextUsageTime) continue;

            maxNextUsageTime = nextUsageTime;
            itemToRemove = item;
        }

        return itemToRemove;
    };

    int answer = 0;

    for (int time = 0; time < K; time++) {
        int curItem = items[time];

        if (connected.contains(curItem)) continue;

        if (connected.size() < N) {
            connected.insert(curItem);
            continue;
        }

        int itemToRemove = getItemToRemove(time);

        connected.erase(itemToRemove);
        connected.insert(curItem);
        answer++;
    }

    cout << answer;

    return 0;
}
```
</details>

저는 다음 제거할 아이템을 매번 찾는 것이 비효율적이라고 생각해서,

아이템 별로 사용 시각을 큐를 사용하여 저장하고,

현재 플러그가 연결된 아이템들의 큐를 확인하면서 가장 나중에 사용될 아이템의 플러그를 제거하는 식으로 구현했습니다.

Munbin-Lee commented:
생각해보니 ```N```, ```K``` 모두 100밖에 되지 않아서 효율을 따지기보다는 문제를 빨리 푸는 게 중요할 수도 있었겠네요.

tgyuuAn commented:
@Munbin-Lee 

오호라 인풋의 사이즈가 커지면 그게 훨씬 빠르겠네요 👍👍
